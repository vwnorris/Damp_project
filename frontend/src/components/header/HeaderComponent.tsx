import AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import Toolbar from '@mui/material/Toolbar';
import Typography from '@mui/material/Typography';
import { Card, Modal, CardContent, Button } from '@mui/material';
import { useState } from 'react';
import ArrowBackIcon from '@mui/icons-material/ArrowBack';
import { Brightness4, Brightness7 } from '@mui/icons-material';

const styleCard = {
  position: 'absolute' ,
  top: '50%',
  left: '50%',
  transform: 'translate(-50%, -50%)',
  width: '80%',
  height: 500,
  overflow: "hidden",
  overflowY: "scroll",
};

export function Header(props: {darkMode: boolean, setDarkMode: (val: boolean) => void}) {
  const [openModalInfo, setOpenModalInfo] = useState(false);

  const handleOpen = () => setOpenModalInfo(true);
  const handleClose = () => setOpenModalInfo(false);

  const refreshPage = () => {
    window.location.reload();
  }

  return (
    <Box sx={{ flexGrow: 1}}>
      <AppBar position='relative' sx={{backgroundColor: (props.darkMode? "##1e1e1e":"#283044")}}>
        <Toolbar sx={{display: 'flex', alignContent: "space-between", width: 1}}>
          <Box sx={{flexGrow: 1}}>
            <Typography variant="h5" onClick={refreshPage} sx={{display: 'inline-block', '&:hover': {
              cursor: 'pointer'
            }}}>
              DAMP
            </Typography>
          </Box>
          <Box>
            <Button color="inherit" onClick={() => {
              localStorage.setItem("dark-mode", JSON.stringify(!props.darkMode))
              props.setDarkMode(!props.darkMode)
              }}
              endIcon={props.darkMode ? <Brightness4/> : <Brightness7/>}
              >
              {props.darkMode ? "Dark" : "Light"}
            </Button>
            <Button color="inherit" onClick={handleOpen}>
              ABOUT
            </Button>
          </Box>
        </Toolbar>
      </AppBar>
      <Modal
        open={openModalInfo}
        onClose={handleClose}
        aria-labelledby="modal-modal-title"
        aria-describedby="modal-modal-description"
        sx={{padding: 4}}>
        <Card sx={{...styleCard}} style={{maxWidth: "650px"}}>
          <CardContent>
              <ArrowBackIcon onClick={handleClose}/>
              <Typography id="modal-modal-title" variant="h5" component="h2" align='center'>
                What is Damp?
              </Typography>
              <Typography id="modal-modal-description" sx={{ mt: 2 }} color="text.primary">
                Damp is a library of games delivered by the popular video game digital distribution service Steam. 
                Today the platform is the largest digital distribution platform for PC games, with over 100 million active users and more than 50 000 games.
                Damp fetches data from the Steam API and displays it in a user-friendly way.
                The user can search for games, see the most important details about them, and write comments to share with other users.
              </Typography>
              <Typography id="modal-modal-title" variant="h5" component="h2" sx={{ mt: 4 }} align='center'>
                How to use Damp?
              </Typography>
              <Typography id="modal-modal-description" sx={{ mt: 2 }} color="text.primary" >
                To get started, you can search for a game by name, developer or publisher.
                In the search results, you can see the game's name, release date, developer, publisher and a brief description.
                You can also see the game's cover image and write a comment about the game.
                In addition, you are able to filter the games on several categories and sort the results of your choice.
              </Typography>
          </CardContent>        
        </Card>
      </Modal>
    </Box>
  );
}
