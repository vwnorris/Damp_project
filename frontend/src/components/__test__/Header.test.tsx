import { Header } from "../header";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import { useEffect, useState } from "react";
import { act } from "react-dom/test-utils";
import renderer from "react-test-renderer";

function TestComponent() {
  const [darkMode, setDarkMode] = useState(false);

  useEffect(() => {
    const val = localStorage.getItem("dark-mode");
    if (val === "true") {
      setDarkMode(true);
    }
  }, [darkMode]);

  return (
    <>
      <p>Is it on? {JSON.stringify(darkMode)}</p>
      <Header darkMode={darkMode} setDarkMode={setDarkMode} />
    </>
  );
}
describe("test header component", () => {
  beforeAll(() => {
    Object.defineProperty(window, 'location', {
      configurable: true,
      value: { reload: jest.fn() },
    });
  });

  it("renders with correct header", () => {
    render(<Header darkMode={false} setDarkMode={() => {}} />);
    screen.getByText(/damp/i);
    screen.getByText(/light/i);
    screen.getByText(/about/i);
  });

  it("test reload of page", () => {
    render(<Header darkMode={false} setDarkMode={() => {}} />);

    act(() => {
      screen.getByText(/damp/i).click()
    })

    screen.getByText(/damp/i)
  });

  it("about loads correctly", async () => {
    render(<Header darkMode={false} setDarkMode={() => {}} />);
    act(() => {
      screen.getByText(/about/i).click();
    })
    act(() => {
      screen.getByText(/light/i).click();
    });
    const cont = await waitFor(() => screen.findByText(/how to use damp/i));

    fireEvent.keyDown(cont, {
      key: "Escape",
      code: "Escape",
      keyCode: 27,
      charCode: 27,
    });

    const modal = screen.queryByText(/how to use damp/i);
    expect(modal).not.toBeInTheDocument();
  });

  it("handles darkmode toggle", async () => {
    localStorage.setItem("dark-mode", "false");
    render(<TestComponent />);
    screen.getByText(/false/i);
    act(() => {
      screen.getByText(/light/i).click();
    });
    await waitFor(() => screen.findByText(/true/i));
    act(() => {
      screen.getByText(/dark/i).click();
    });
    await waitFor(() => screen.findByText(/false/i));
  });
});

describe("header snapshot tests", () => {
  it("basic snapshot test", () => {
    const tree = renderer
      .create(<Header darkMode={false} setDarkMode={() => {}} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
