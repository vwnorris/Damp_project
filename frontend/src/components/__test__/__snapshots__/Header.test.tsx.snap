// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`header snapshot tests basic snapshot test 1`] = `
<div
  className="MuiBox-root css-i9gxme"
>
  <header
    className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionRelative css-ep400p-MuiPaper-root-MuiAppBar-root"
  >
    <div
      className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-mr2j8p-MuiToolbar-root"
    >
      <div
        className="MuiBox-root css-i9gxme"
      >
        <h5
          className="MuiTypography-root MuiTypography-h5 css-14bwlsx-MuiTypography-root"
          onClick={[Function]}
        >
          DAMP
        </h5>
      </div>
      <div
        className="MuiBox-root css-0"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit css-1y942vo-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Light
          <span
            className="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="Brightness7Icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69zM12 18c-3.31 0-6-2.69-6-6s2.69-6 6-6 6 2.69 6 6-2.69 6-6 6zm0-10c-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4z"
              />
            </svg>
          </span>
        </button>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit css-1y942vo-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          ABOUT
        </button>
      </div>
    </div>
  </header>
</div>
`;
